Examples/WindowsFormsApplication/Nfield.cs

```diff
@@ -0,0 +1,138 @@
+        {
+            var client = PublicClientApplicationBuilder.Create(ApplicationConfiguration.Current.ClientId)
+                .WithRedirectUri("http://localhost")
+                .WithAuthority(AzureCloudInstance.AzurePublic, ApplicationConfiguration.Current.Tenant)
+                .Build();
+
+            client.UserTokenCache.SetBeforeAccess(BeforeAccessNotification);
+            client.UserTokenCache.SetAfterAccess(AfterAccessNotification);
+
+            return client;
+        }
+
+        private async Task<string> ProvideTokenAsync()
+        {
+            var result = await AuthenticateAsync((status) => { });
+            return result.AccessToken;
+        }
+
+        private static string CreateScope(string scope) => $"{ApplicationConfiguration.Current.NfieldApiApplicationId}/{scope}";
+
+        internal async Task LogoutAsync()
+        {
+            var accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            while (accounts.Any())
+            {
+                await ClientApp.RemoveAsync(accounts.First());
+                accounts = (await ClientApp.GetAccountsAsync()).ToList();
+            }
+        }
+
+        private byte[] RestoreTokenCacheFromRegistry()
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                var value = key.GetValue("TokenCache");
+                if (value is byte[] state)
+                    return state;
+            }
+
+            return null;
+        }
+
+        private void SaveTokenCacheInRegistry(byte[] state)
+        {
+            using (RegistryKey key = Registry.CurrentUser.CreateSubKey(@"SOFTWARE\NIPO Samples"))
+            {
+                key.SetValue("TokenCache", state, RegistryValueKind.Binary);
+            }
+        }
+
+        public void BeforeAccessNotification(TokenCacheNotificationArgs args)
+        {
+            args.TokenCache.DeserializeMsalV3(RestoreTokenCacheFromRegistry());
+        }
+
+        public void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            if (args.HasStateChanged)
+            {
+                SaveTokenCacheInRegistry(args.TokenCache.SerializeMsalV3());
+            }
+        }
+
+        public async Task<AuthenticationResult> TryAuthenticateSilentAsync()
+        {
+            var accounts = await ClientApp.GetAccountsAsync();
+            var account = accounts.FirstOrDefault();
+
+            try
+            {
+                return await ClientApp.AcquireTokenSilent(Scopes, account).ExecuteAsync();
+            }
+            catch (MsalUiRequiredException)
+            {
+                return null;
+            }
+        }
+
+        public async Task<AuthenticationResult> AuthenticateAsync(Action<string> statusCallback)
+        {
+            var accounts = await ClientApp.GetAccountsAsync();
+            var account = accounts.FirstOrDefault();
```

Comment:
Is it encouraged to always take the first account? why? in logout i saw also an enumeration of multipe accounts. I am not super familiar with this new libs yet so I wonder if this is good practice :-).

Reply:
It's what they use in all the sample. This library also supports multiple account, but this app only ever uses one, however the API is geared towards multiple.